File: test/scripts/blocking-io.py
#!/usr/bin/env python3
"""
ANTI-PATTERN: Blocking I/O instead of async
Resume: "Use Asynchronous I/O" - Blocking ties up threads; with many
concurrent ops you need many threads or poor scalability.

Run: python blocking-io.py
(Simulates blocking by doing sync file reads in threads)
"""

import concurrent.futures
import tempfile
import time
import os

NUM_READS = 100
FILE_SIZE_KB = 100


def main():
    print("ANTI-PATTERN: Blocking I/O (sync in threads)")
    print(f"{NUM_READS} concurrent sync file reads")

    with tempfile.NamedTemporaryFile(delete=False, suffix='.dat') as f:
        f.write(os.urandom(FILE_SIZE_KB * 1024))
        path = f.name

    try:
        def blocking_read():
            with open(path, 'rb') as f:
                return len(f.read(FILE_SIZE_KB * 1024))

        # BAD: Many threads doing blocking I/O
        start = time.perf_counter()
        with concurrent.futures.ThreadPoolExecutor(max_workers=NUM_READS) as ex:
            list(ex.map(lambda _: blocking_read(), range(NUM_READS)))
        elapsed = time.perf_counter() - start

        print(f"Blocking (many threads): {elapsed*1000:.0f} ms")
        print("Fix: use asyncio + aiofiles or ThreadPool for I/O-bound")
    finally:
        os.unlink(path)


if __name__ == "__main__":
    main()

File: test/scripts/blocking-io.js
#!/usr/bin/env node
/**
 * ANTI-PATTERN: Blocking I/O (sync) instead of async
 * Resume: "Use Asynchronous I/O" - Sync blocks the event loop.
 *
 * Run: node blocking-io.js
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const NUM_READS = 100;
const FILE_SIZE_KB = 100;

async function main() {
  const tmp = path.join(os.tmpdir(), `perf-blocking-io-${Date.now()}.dat`);
  fs.writeFileSync(tmp, Buffer.alloc(FILE_SIZE_KB * 1024));

  console.log('ANTI-PATTERN: Blocking I/O (sync)');
  console.log(`${NUM_READS} sync file reads (blocks event loop)`);

  const start = Date.now();
  for (let i = 0; i < NUM_READS; i++) {
    fs.readFileSync(tmp);
  }
  const elapsed = Date.now() - start;

  fs.unlinkSync(tmp);
  console.log(`Sync reads: ${elapsed} ms`);
  console.log('Fix: use fs.promises.readFile or fs.readFile with callback');
}

main().catch(console.error);
